File: src/client/java/fr/hardel/asset_editor/client/javafx/components/page/enchantment/EnchantmentTreeBuilder.java
package fr.hardel.asset_editor.client.javafx.components.page.enchantment;

import fr.hardel.asset_editor.client.javafx.components.ui.tree.TreeNodeModel;
import fr.hardel.asset_editor.client.javafx.lib.data.EnchantmentTreeData;
import fr.hardel.asset_editor.client.javafx.lib.data.EnchantmentViewMatchers;
import fr.hardel.asset_editor.client.javafx.lib.data.SlotConfigs;
import fr.hardel.asset_editor.client.javafx.lib.data.SlotConfigs.SlotConfig;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioSidebarView;
import fr.hardel.asset_editor.client.javafx.lib.text.StudioText;
import fr.hardel.asset_editor.client.javafx.lib.store.RegistryElementStore.ElementEntry;
import net.minecraft.resources.Identifier;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class EnchantmentTreeBuilder {

    public static TreeNodeModel build(Collection<ElementEntry<?>> entries, StudioSidebarView view) {
        List<ElementEntry<Enchantment>> enchantments = castEntries(entries);

        TreeNodeModel root = new TreeNodeModel();
        root.setCount(enchantments.size());
        if (view == StudioSidebarView.SLOTS) {
            buildBySlots(root, enchantments);
            return root;
        }
        if (view == StudioSidebarView.ITEMS) {
            buildByItems(root, enchantments);
            return root;
        }
        buildByExclusive(root, enchantments);
        return root;
    }

    public static Map<String, Identifier> slotFolderIcons() {
        LinkedHashMap<String, Identifier> icons = new LinkedHashMap<>();
        for (SlotConfig config : SlotConfigs.ALL) icons.put(config.id(), config.image());
        return icons;
    }

    public static Map<String, Identifier> itemFolderIcons() {
        LinkedHashMap<String, Identifier> icons = new LinkedHashMap<>();
        for (EnchantmentTreeData.ItemTagConfig tag : EnchantmentTreeData.ITEM_TAGS) {
            icons.put(tag.key(), tag.icon());
        }
        return icons;
    }

    private static void buildBySlots(TreeNodeModel root, List<ElementEntry<Enchantment>> enchantments) {
        for (SlotConfig config : SlotConfigs.ALL) {
            List<ElementEntry<Enchantment>> matching = enchantments.stream()
                    .filter(e -> EnchantmentViewMatchers.matchesSlot(e, config.id()))
                    .toList();
            if (!matching.isEmpty()) {
                TreeNodeModel category = createCategoryNode(matching);
                category.setIcon(config.image());
                category.setLabel(StudioText.resolve(StudioText.Domain.SLOT, config.id()));
                root.children().put(config.id(), category);
            }
        }
    }

    private static void buildByItems(TreeNodeModel root, List<ElementEntry<Enchantment>> enchantments) {
        for (EnchantmentTreeData.ItemTagConfig tag : EnchantmentTreeData.ITEM_TAGS) {
            List<ElementEntry<Enchantment>> matching = enchantments.stream()
                    .filter(e -> EnchantmentViewMatchers.matchesItem(e, tag.key()))
                    .toList();
            if (!matching.isEmpty()) {
                TreeNodeModel category = createCategoryNode(matching);
                category.setIcon(tag.icon());
                category.setLabel(StudioText.resolve(StudioText.Domain.ENCHANTMENT_SUPPORTED, tag.key()));
                root.children().put(tag.key(), category);
            }
        }
    }

    private static void buildByExclusive(TreeNodeModel root, List<ElementEntry<Enchantment>> enchantments) {
        LinkedHashMap<String, List<ElementEntry<Enchantment>>> grouped = new LinkedHashMap<>();
        for (var entry : enchantments) {
            var tagKey = entry.data().exclusiveSet().unwrapKey();
            if (tagKey.isPresent()) {
                grouped.computeIfAbsent(tagKey.get().location().getPath(), k -> new ArrayList<>()).add(entry);
                continue;
            }

            List<Identifier> directIds = entry.data().exclusiveSet().stream()
                    .map(holder -> holder.unwrapKey().map(k -> k.identifier()).orElse(null))
                    .filter(Objects::nonNull)
                    .toList();
            if (directIds.isEmpty()) {
                continue;
            }
            for (Identifier id : directIds) {
                grouped.computeIfAbsent(id.toString(), k -> new ArrayList<>()).add(entry);
            }
        }
        for (var e : grouped.entrySet()) {
            TreeNodeModel category = createCategoryNode(e.getValue());
            Identifier tagId = Identifier.tryParse(e.getKey());
            String setName = tagId == null ? e.getKey()
                    : tagId.getPath().substring(tagId.getPath().lastIndexOf('/') + 1);
            category.setLabel(StudioText.resolve(StudioText.Domain.ENCHANTMENT_EXCLUSIVE, setName));
            root.children().put(e.getKey(), category);
        }
    }

    private static TreeNodeModel createCategoryNode(List<ElementEntry<Enchantment>> enchantments) {
        TreeNodeModel node = new TreeNodeModel();
        node.setCount(enchantments.size());
        for (var entry : enchantments) {
            TreeNodeModel leaf = new TreeNodeModel();
            leaf.setCount(0);
            leaf.setElementId(entry.id().toString());
            leaf.setLabel(entry.data().description().getString());
            node.children().put(entry.id().getPath(), leaf);
        }
        return node;
    }

    @SuppressWarnings("unchecked")
    private static List<ElementEntry<Enchantment>> castEntries(Collection<ElementEntry<?>> entries) {
        return entries.stream().map(e -> (ElementEntry<Enchantment>) e).toList();
    }

    private EnchantmentTreeBuilder() {}
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/page/enchantment/EnchantmentTreeBuilder.kt
package fr.hardel.asset_editor.client.compose.components.page.enchantment

import fr.hardel.asset_editor.client.compose.components.ui.tree.TreeNodeModel
import fr.hardel.asset_editor.client.compose.lib.StudioText
import fr.hardel.asset_editor.client.compose.lib.SlotConfigs
import fr.hardel.asset_editor.workspace.ElementEntry
import java.util.Locale
import net.minecraft.client.resources.language.I18n
import net.minecraft.resources.Identifier
import net.minecraft.world.item.enchantment.Enchantment

object EnchantmentTreeBuilder {

    fun build(
        enchantments: List<ElementEntry<Enchantment>>,
        view: StudioSidebarView
    ): TreeNodeModel {
        val root = TreeNodeModel()
        root.count = enchantments.size

        when (view) {
            StudioSidebarView.SLOTS -> buildBySlots(root, enchantments)
            StudioSidebarView.ITEMS -> buildByItems(root, enchantments)
            StudioSidebarView.EXCLUSIVE -> buildByExclusive(root, enchantments)
        }

        return root
    }

    fun slotFolderIcons(): Map<String, Identifier> =
        LinkedHashMap<String, Identifier>().apply {
            SlotConfigs.PHYSICAL.forEach { id -> put(id, SlotConfigs.slotImage(id)) }
        }

    fun itemFolderIcons(): Map<String, Identifier> =
        LinkedHashMap<String, Identifier>().apply {
            EnchantmentTreeData.ITEM_TAGS.forEach { tag -> put(tag.key(), tag.icon()) }
        }

    private fun buildBySlots(root: TreeNodeModel, enchantments: List<ElementEntry<Enchantment>>) {
        SlotConfigs.PHYSICAL.forEach { slotId ->
            val matching = enchantments.filter { entry -> EnchantmentViewMatchers.matchesSlot(entry, slotId) }
            if (matching.isEmpty()) {
                return@forEach
            }

            val category = createCategoryNode(matching)
            category.icon = SlotConfigs.slotImage(slotId)
            category.label = I18n.get("slot:$slotId")
            root.children[slotId] = category
        }
    }

    private fun buildByItems(root: TreeNodeModel, enchantments: List<ElementEntry<Enchantment>>) {
        EnchantmentTreeData.ITEM_TAGS.forEach { tag ->
            val matching = enchantments.filter { entry -> EnchantmentViewMatchers.matchesItem(entry, tag.key()) }
            if (matching.isEmpty()) {
                return@forEach
            }

            val category = createCategoryNode(matching)
            category.icon = tag.icon()
            category.label = StudioText.resolve("item_tag", tag.tagId)
            root.children[tag.key()] = category
        }
    }

    private fun buildByExclusive(root: TreeNodeModel, enchantments: List<ElementEntry<Enchantment>>) {
        val grouped = LinkedHashMap<String, MutableList<ElementEntry<Enchantment>>>()

        enchantments.forEach { entry ->
            val tagKey = entry.data().exclusiveSet().unwrapKey()
            if (tagKey.isPresent) {
                grouped.computeIfAbsent(tagKey.get().location().path) { ArrayList() }.add(entry)
                return@forEach
            }

            val directIds = entry.data().exclusiveSet().stream()
                .map { holder -> holder.unwrapKey().map { key -> key.identifier() }.orElse(null) }
                .filter { id -> id != null }
                .toList()

            directIds.forEach { identifier ->
                grouped.computeIfAbsent(identifier.toString()) { ArrayList() }.add(entry)
            }
        }

        grouped.forEach { (key, value) ->
            val category = createCategoryNode(value)
            val tagId = Identifier.tryParse(key)
            category.label = if (tagId != null) {
                StudioText.resolve("enchantment_tag", tagId)
            } else {
                key.lowercase(Locale.ROOT)
            }
            root.children[key] = category
        }
    }

    private fun createCategoryNode(enchantments: List<ElementEntry<Enchantment>>): TreeNodeModel {
        val node = TreeNodeModel()
        node.count = enchantments.size

        enchantments.forEach { entry ->
            val leaf = TreeNodeModel()
            leaf.elementId = entry.id().toString()
            leaf.label = entry.data().description().string
            node.children[entry.id().toString()] = leaf
        }

        return node
    }
}
